### PR TITLE
External Opensearch Support

### PR DESCRIPTION
**Pull request by jashaik**
### Description

1. Migrate External Elasticsearch to External Openserach in Chef Infra Server to 1.2
2. This PR suppose to support only external opensearch.

### Issues Resolved

https://app.zenhub.com/workspaces/chef-infra-server-team-5fc64867d45ca500173dbbc7/issues/chef/chef-server/3068
https://chefio.atlassian.net/browse/INFS-38
https://chefio.atlassian.net/browse/INFS-26
https://chefio.atlassian.net/browse/INFS-2

### Check List

- [ ] New functionality includes tests
- [ ] All buildkite tests pass
- [ ] Full omnibus build and tests in buildkite pass
- [ ] All commits have been signed-off for the Developer Certificate of Origin. See <https://github.com/chef/chef/blob/main/CONTRIBUTING.md#developer-certification-of-origin-dco>
- [ ] PR title is a worthy inclusion in the CHANGELOG

### Questions
- There many seach related attributes in the config file, Should be optimized?
   Discussed - Follow the elasticsearch way of configurations for now.
- Can we use the given opensearch auth credentials for erchef operations ?
  Discussed - Yes, Use the user given auth credentials for external Opensearch auth.
- Identify the major & minor version bumps in opensearch.
   AC - shahid
- Can we use existing way of version checks in the erchef erlang modules ?
  Discussed - Need to use serach_providers in the chef_index to make a call to chef_elasticsearch or chef_opensearch
`case {serach_providers, envy:get(chef_index, solr_elasticsearch_major_version, 2, non_neg_integer)} of
        X when X==7; X==1 ->
            IndexEjson;
        _ ->
            [{<<"_type">>, <<"object">>}| IndexEjson]
    end,`
### Assumptions

1.  Migrations need reindexing of openserach.
2. Or Follow the steps here for migrating the data from Elasticsearch to Opensearch - https://opensearch.org/docs/latest/upgrade-to/upgrade-to/ for migration

### CODE REVIEW

File: omnibus/files/server-ctl-cookbooks/infra-server/resources/opensearch_index.rb [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+action :destroy do
+  if index_exists?
+    converge_by "Deleting elasticsearch index #{new_resource.index_name}" do
+      search_engine_server.delete(new_resource.index_name, search_engine_auth_header)
+    end
+  end
+end
+
+action_class do
+  def create_opensearch_user
```
Comment (jashaik) [inline omnibus/files/server-ctl-cookbooks/infra-server/resources/opensearch_index.rb:48]: Add description

File: src/oc_erchef/apps/chef_index/src/chef_elasticsearch.erl [hunk 2]
```diff
@@ -133,15 +133,15 @@ query_body(#chef_solr_query{
...
     case envy:get(chef_index, solr_elasticsearch_major_version, 2, non_neg_integer) of
-        X when X >= 5 -> <<"stored_fields">>;
+        X when X >= 5; X==1 -> <<"stored_fields">>;
         2 -> <<"fields">>
     end.
 
 query_string_query_ejson(QueryString) ->
     QueryEjson = [{<<"query">>, list_to_binary(QueryString)}],
     QueryEjson1 =
     case envy:get(chef_index, solr_elasticsearch_major_version, 2, non_neg_integer) of
```
Comment (jashaik) [inline src/oc_erchef/apps/chef_index/src/chef_elasticsearch.erl:143]: {search_provider, version}

File: oc-chef-pedant/lib/pedant/rspec/search_util.rb [hunk 3]
```diff
@@ -708,9 +708,11 @@ def direct_solr_query(type, query)
         # Chef Object type (in addition to filtering based on org).  We
         # don't have an easy way to access the org's guid in the tests, so
         # I'm not using that. In any event, the following query works.
+        auth = Base64.strict_encode64("admin:admin")
         url = "#{Pedant::Config.search_server}#{Pedant::Config.search_url_fmt}" % {:type => CGI.escape(type), :query => CGI.escape(query)}
         headers = {
-          "Accept" => "application/json"
+          "Accept" => "application/json",
+          "Authorization" => "Basic #{auth}"
```
Comment (PrajaktaPurohit) [inline oc-chef-pedant/lib/pedant/rspec/search_util.rb:715]: The additional headers will not work for elasticsearch, correct? They should be included only if the provider is elasticsearch.

Comment (PrajaktaPurohit) [reply oc-chef-pedant/lib/pedant/rspec/search_util.rb:715]: we probably need something like maybe_add_headers(), perhaps in the helper that will add the Auth header for opensearch search provider only.

**Comment by jashaik** [comment]
Pipeline Information -
Adhoc - https://buildkite.com/chef/chef-chef-server-main-omnibus-adhoc/builds/3809
Umreblla - https://buildkite.com/chef/chef-umbrella-main-chef-server/builds/1026

**Comment by jashaik** [comment]
Testing status -

- All the pedant tests are passing except stats_spec.rb as the counters are different for opensearch so planned to use `search_provider` but not allowed to use inside the describe block.
- Unit tests -> Vinay is working on them
- Document changes -> Shahid is working.
- Latest pipelines - 
cc - @PrajaktaPurohit @marcparadise @vinay-satish @lbakerchef

File: oc-chef-pedant/lib/pedant/rspec/search_util.rb [hunk 4]
```diff
@@ -697,6 +697,24 @@ def search_result_databag_item(bag_name, item_name, item)
         }
       end
 
+      # Added for opensearch authorization header
+      def get_search_headers()
+        config = ::JSON.parse(IO.read("/etc/opscode/chef-server-running.json"))['private_chef']
+        if config['opscode-erchef']['search_provider'] == 'opensearch'
+          username = config['opscode-erchef']['search_auth_username']
+          password = config['opscode-erchef']['search_auth_username']
```
Comment (jashaik) [inline oc-chef-pedant/lib/pedant/rspec/search_util.rb:705]: Fix it

File: omnibus/files/server-ctl-cookbooks/infra-server/attributes/default.rb [hunk 5]
```diff
@@ -319,8 +335,8 @@
 default['private_chef']['opscode-erchef']['search_batch_max_size'] = '5000000'
 default['private_chef']['opscode-erchef']['search_batch_max_wait'] = '10'
 default['private_chef']['opscode-erchef']['search_auth_enabled'] = false
-default['private_chef']['opscode-erchef']['search_auth_username'] = 'admin'
-default['private_chef']['opscode-erchef']['search_auth_password'] = 'admin'
+default['private_chef']['opscode-erchef']['search_auth_username'] = "admin"
+default['private_chef']['opscode-erchef']['search_auth_password'] = "admin"
```
Comment (PrajaktaPurohit) [inline omnibus/files/server-ctl-cookbooks/infra-server/attributes/default.rb:339]: the default password can probably be something else?

Comment (PrajaktaPurohit) [reply omnibus/files/server-ctl-cookbooks/infra-server/attributes/default.rb:339]: check if these belong under opscode-erchef or opensearch hash.

File: omnibus/files/server-ctl-cookbooks/infra-server/libraries/helper.rb [hunk 6]
```diff
@@ -55,51 +55,155 @@ def vip_for_uri(service)
...
+      # variable but, there was a support to give the external elasticsearch url in the 
+      # node['private_chef']['opscode-solr4']['external_url'] variable. We still need to support this.
+      node['private_chef']['opscode-solr4']['external']
+    else
+      false
+    end
+  end
+
+  def search_engine_url
+    res = node['private_chef'][search_provider]['external_url'] || node['private_chef']['opscode-solr4']['external_url']
```
Comment (jashaik) [inline omnibus/files/server-ctl-cookbooks/infra-server/libraries/helper.rb:81]: remove

File: omnibus/files/server-ctl-cookbooks/infra-server/libraries/helper.rb [hunk 7]
```diff
@@ -55,51 +55,155 @@ def vip_for_uri(service)
...
+    if external_search_engine?
+      # Ideally the external elasticsearch url should be in the node['private_chef']['elasticsearch']['external']
+      # variable but, there was a support to give the external elasticsearch url in the 
+      # node['private_chef']['opscode-solr4']['external_url'] variable. We still need to support this.
+      node['private_chef'][search_provider]['external_url'] || node['private_chef']['opscode-solr4']['external_url']
```
Comment (jashaik) [inline omnibus/files/server-ctl-cookbooks/infra-server/libraries/helper.rb:86]: Note: Opensearch - user opensearch attribute

File: omnibus/files/server-ctl-cookbooks/infra-server/libraries/helper.rb [hunk 8]
```diff
@@ -55,51 +55,155 @@ def vip_for_uri(service)
...
+      end
+      if (search_provider == 'opensearch') && !([1].include?(version))
+        raise "Unsupported Opensearch version of #{version}. There is currently support for the major versions of 1."
       end
-      raise "Unsupported Elasticsearch version of #{version}. There is currently support for the major versions of 2, 5, 6 and 7." unless [2, 5, 6, 7].include?(version)
       version
     else
-      # Elasticsearch is disabled - this configuration setting should never be used in erlang.
+      # opscode-solr4 is enabled - this configuration setting should never be used in erlang.
       0
```
Comment (jashaik) [inline omnibus/files/server-ctl-cookbooks/infra-server/libraries/helper.rb:140]: Verify: check if it is part of preflight checks

File: omnibus/files/server-ctl-cookbooks/infra-server/libraries/helper.rb [hunk 9]
```diff
@@ -445,6 +532,7 @@ def self.chef_server_running_content(node)
     # back-compat fixes for opscode-reporting opscode-reporting still
     # won't work with non-external Elasticsearch until reporting is
     # updated.
+    # ToDo - Should we use search_provider for this
```
Comment (jashaik) [inline omnibus/files/server-ctl-cookbooks/infra-server/libraries/helper.rb:535]: Remove

File: omnibus/files/server-ctl-cookbooks/infra-server/recipes/default.rb [hunk 10]
```diff
@@ -148,7 +148,11 @@
       action :disable
     end
   elsif node['private_chef'][service]['enable']
-    include_recipe "infra-server::#{service}"
+    if service == 'elasticsearch' && node['private_chef']['opscode-erchef']['search_provider'] == 'opensearch' && node['private_chef']['opensearch']['external']
```
Comment (jashaik) [inline omnibus/files/server-ctl-cookbooks/infra-server/recipes/default.rb:151]: Preflight check: If the search_provider is opensearch & opensearch external is not enabled then raise it in preflight checks

Comment (jashaik) [reply omnibus/files/server-ctl-cookbooks/infra-server/recipes/default.rb:151]: Add opensearch-external.rb

File: omnibus/files/server-ctl-cookbooks/infra-server/resources/opensearch_index.rb [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+action :destroy do
+  if index_exists?
+    converge_by "Deleting elasticsearch index #{new_resource.index_name}" do
+      search_engine_server.delete(new_resource.index_name, new_resource.opensearch_auth)
+    end
+  end
+end
+
+action_class do
+  # Tobe removed after review
```
Comment (marcparadise) [inline omnibus/files/server-ctl-cookbooks/infra-server/resources/opensearch_index.rb:48]: This looks like the right thing to do, so I'm not sure why we're not.  Could you provide some context?

File: omnibus/files/server-ctl-cookbooks/infra-server/recipes/default.rb (hunk 10 above)
Comment (marcparadise) [reply omnibus/files/server-ctl-cookbooks/infra-server/recipes/default.rb:151]: This still seems off to me. I know we talked about it, but could you refresh my memory - why can't we stick to the basic "infra-server::#{service}" and "infra-server::#{service-external}" behavior here?

**Review by marcparadise — CHANGES_REQUESTED** [review]

File: omnibus/files/server-ctl-cookbooks/infra-server/attributes/default.rb (hunk 5 above)
Comment (jashaik) [reply omnibus/files/server-ctl-cookbooks/infra-server/attributes/default.rb:339]: For external and internal, We expect the username & password for connecting to Opensearch.
External Scenario - We will not have more control on their opensearch setup(like updating the security config files for adding the secrets generated username & password).So like postgres We except the credentials from the customer.
Internal Scenario - We will be having the access to opensearch config files so we can add the generated username & password in the Security configs.
In short, Credantials should be passed in the external case but not in the internal.

Yes, We don't need these defaults set in the config file.And Preflight checks error out if they set search_provider to opensearch without setting 'search_auth_username' & 'search_auth_password'.

File: omnibus/files/server-ctl-cookbooks/infra-server/libraries/preflight_opensearch_validator.rb [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+    @cs_opensearch_attr = PrivateChef['opensearch']
+    @helper = OmnibusHelper.new(node)
+    @cs_erchef_attr = PrivateChef['opscode_erchef']
+
+    @node_opensearch_attr = node['private_chef']['opensearch']
+    @node_erchef_attr = node['private_chef']['opscode-erchef']
+  end
+
+  def run!
+    verify_system_memory
```
[the remark this answers is not in the document]
Comment (jashaik) [reply omnibus/files/server-ctl-cookbooks/infra-server/libraries/preflight_opensearch_validator.rb:46]: Yes, It is not relevant for external. I will skip it If it is external.

File: omnibus/files/server-ctl-cookbooks/infra-server/libraries/preflight_opensearch_validator.rb [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+
+  def verify_external_url
+    if cs_opensearch_attr['external'] && !cs_opensearch_attr['external_url']
+      fail_with err_INDEX006_bad_external_config()
+    end
+  end
+
+  def verify_erchef_config
+    provider = @cs_erchef_attr['search_provider']
+    return true if provider.nil? # default provider
```
[the remark this answers is not in the document]
Comment (jashaik) [reply omnibus/files/server-ctl-cookbooks/infra-server/libraries/preflight_opensearch_validator.rb:168]: Currently We made this as a mandatory field but in future we can remove with 'search_provider' attribute.

File: omnibus/files/server-ctl-cookbooks/infra-server/libraries/preflight_opensearch_validator.rb [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+
+    unless %w(batch inline).include?(@cs_erchef_attr['search_queue_mode'])
+      fail_with err_INDEX007_bad_queue_mode
+    end
+  end
+
+  def err_INDEX001_failed_validation(final_min, final_max)
```
[the remark this answers is not in the document]
Comment (jashaik) [reply omnibus/files/server-ctl-cookbooks/infra-server/libraries/preflight_opensearch_validator.rb:175]: I am planning to use `OPENSEARCH001`.

File: omnibus/files/server-ctl-cookbooks/infra-server/resources/opensearch_index.rb (hunk 11 above)
Comment (jashaik) [reply omnibus/files/server-ctl-cookbooks/infra-server/resources/opensearch_index.rb:48]: In the Internal scenario, The user creation should be by updating the security config file not with the API call for fresh install.
In the External scenario, We can't have the full control so expecting the credentials from them and using the same in Erchef application.
Hence I commented out this creating funtion.I will be removing this code block.

File: docs-chef-io/content/server/upgrades.md [hunk 15]
```diff
@@ -46,6 +46,16 @@ Chef Infra Server 14.0 moved from Solr to Elasticsearch as its search index.
 
 The Chef Infra Server 14 upgrade does not automatically reindex existing external Elasticsearch installations.
 
+#### Upgrading to 14.13
+
+Chef Infra Server 14.13 supports External Opensearch for indexing. Migrating from Elasticsearch to External Opensearch needs ontime reindexing and reconfigure after upgrade. This process may take longer depending on your server hardware and the size of the node objects on your Chef Infra Server.
+
```
Comment (PrajaktaPurohit) [inline docs-chef-io/content/server/upgrades.md:52]: https://docs.chef.io/server/upgrades/#standalone-server useful to give some estimates. Even better if we could provide what the estimates are based off of.

File: oc-chef-pedant/pedant_config.rb [hunk 16]
```diff
@@ -68,6 +68,7 @@
 # poll for a period to ensure they are querying committed results.
 search_server "http://127.0.0.1:8983"
 
+
```
Comment (PrajaktaPurohit) [inline oc-chef-pedant/pedant_config.rb:71]: [NIT] Extra space?

**Review by marcparadise — APPROVED** [review]

File: docs-chef-io/content/server/upgrades.md [hunk 17]
```diff
@@ -46,6 +46,16 @@ Chef Infra Server 14.0 moved from Solr to Elasticsearch as its search index.
 
 The Chef Infra Server 14 upgrade does not automatically reindex existing external Elasticsearch installations.
 
+#### Upgrading to 14.13
+
+Chef Infra Server 14.13 supports External Opensearch for indexing. Migrating from Elasticsearch to External Opensearch needs ontime reindexing and reconfigure after upgrade. This process may take longer depending on your server hardware and the size of the node objects on your Chef Infra Server.
```
Comment (kagarmoe) [inline docs-chef-io/content/server/upgrades.md:51]:
```suggestion
Chef Infra Server 14.13 supports External Opensearch for indexing. Migrating from Elasticsearch to External Opensearch requires reindexing and reconfiguring after the upgrade. This process may take longer depending on your server hardware and the size of the node objects on your Chef Infra Server.
```

File: docs-chef-io/content/server/upgrades.md [hunk 18]
```diff
@@ -46,6 +46,16 @@ Chef Infra Server 14.0 moved from Solr to Elasticsearch as its search index.
...
+
+{{< note >}}
+
+Set the `opensearch['external']` & `opensearch['external_url']` attribute in [chef-server.rb]({{< relref "config_rb_server_optional_settings" >}}) to enable external Opensearch for indexing.
```
Comment (kagarmoe) [inline docs-chef-io/content/server/upgrades.md:55]: Since this setting is required, it should be presented as a step instead of in a note. Please add example values for the config settings.

**Review by kagarmoe — CHANGES_REQUESTED** [review]
Left comments for a few changes

File: docs-chef-io/content/server/upgrades.md (hunk 18 above)
Comment (jashaik) [reply docs-chef-io/content/server/upgrades.md]: Updated as per the feedback. Please re-verify

File: docs-chef-io/content/server/upgrades.md [hunk 19]
```diff
@@ -46,6 +46,18 @@ Chef Infra Server 14.0 moved from Solr to Elasticsearch as its search index.
...
+
+Chef Infra Server 14.13 supports External Opensearch for indexing. Migrating from Elasticsearch to External Opensearch requires reindexing and reconfiguring after the upgrade. This process may take longer depending on your server hardware and the size of the node objects on your Chef Infra Server.
+
+#### Steps to enable external OpenSearch
+1. Set the `opensearch['external']` attribute to `true`
+2. Set the `opensearch['external_url']` attribute to external OpenSearch url. Example: `http://127.0.0.1:9200`.
+   ```bash
+      opensearch['external'] = true
+      opensearch['external_url'] = "http://127.0.0.1:9200"
+   ```
```
Comment (IanMadd) [inline docs-chef-io/content/server/upgrades.md:59]:
````suggestion
#### Steps To Enable External OpenSearch

1. Set the `opensearch['external']` attribute to `true`.
1. Set the `opensearch['external_url']` attribute to the external OpenSearch URL.

For example:

```bash
opensearch['external'] = true
opensearch['external_url'] = "http://127.0.0.1:9200"
```
````

File: docs-chef-io/content/server/v14/config_rb_server_optional_settings.md [hunk 20]
```diff
@@ -1472,6 +1472,22 @@ Be careful if increasing this number - requests for a given set of cookbook vers
 not occur if you increment the version of a cookbook with every cookbook update, which is the recommended approach to updating cookbooks.
 {{< /note >}}
 
+### OpenSearch
+
+This configuration file has the following settings for `opensearch`:
+
+`opensearch['external']`
+
+: Enable external `opensearch` service by setting `true`. Default value: `false`.
```
Comment (IanMadd) [inline docs-chef-io/content/server/v14/config_rb_server_optional_settings.md:1481]:
```suggestion
: Enable external `opensearch` service by setting to `true`. Default value: `false`.
```

File: docs-chef-io/content/server/v14/config_rb_server_optional_settings.md [hunk 21]
```diff
@@ -1472,6 +1472,22 @@ Be careful if increasing this number - requests for a given set of cookbook vers
...
+
+`opensearch['external_url']`
+
+: External opensearch url. Example: `http://127.0.0.1:9200`. Default value: `nil`
```
Comment (IanMadd) [inline docs-chef-io/content/server/v14/config_rb_server_optional_settings.md:1485]:
```suggestion
: The external OpenSearch URL. Example: `http://127.0.0.1:9200`. Default value: `nil`
```

File: docs-chef-io/content/server/v14/config_rb_server_optional_settings.md [hunk 22]
```diff
@@ -1472,6 +1472,22 @@ Be careful if increasing this number - requests for a given set of cookbook vers
...
+
+{{< note >}}
+Chef Infra Server supports OpenSearch only as an external indexing provider. Both `external` and `external_url` must be provided under this configuration.
```
Comment (IanMadd) [inline docs-chef-io/content/server/v14/config_rb_server_optional_settings.md:1488]:
```suggestion
Chef Infra Server supports OpenSearch only as an external indexing provider. You must provide values for `external` and `external_url` under this configuration.
```

File: docs-chef-io/content/server/upgrades.md [hunk 23]
```diff
@@ -46,6 +46,18 @@ Chef Infra Server 14.0 moved from Solr to Elasticsearch as its search index.
 
 The Chef Infra Server 14 upgrade does not automatically reindex existing external Elasticsearch installations.
 
+#### Upgrading to 14.13
+
+Chef Infra Server 14.13 supports External Opensearch for indexing. Migrating from Elasticsearch to External Opensearch requires reindexing and reconfiguring after the upgrade. This process may take longer depending on your server hardware and the size of the node objects on your Chef Infra Server.
```
Comment (IanMadd) [inline docs-chef-io/content/server/upgrades.md:51]:
```suggestion
Chef Infra Server 14.13 supports external OpenSearch for indexing. To migrate from Elasticsearch to external OpenSearch, you must reindex and reconfigure your database after upgrading to Chef Infra Server 14.13. The duration of this operation will vary depending on your server hardware and the number of node objects on your Chef Infra Server.
```

**Review by IanMadd — COMMENTED** [review]
Docs feedback

File: docs-chef-io/content/server/upgrades.md [hunk 24]
```diff
@@ -46,6 +46,19 @@ Chef Infra Server 14.0 moved from Solr to Elasticsearch as its search index.
...
+Chef Infra Server 14.13 supports external OpenSearch for indexing. To migrate from Elasticsearch to external OpenSearch, you must reindex and reconfigure your database after upgrading to Chef Infra Server 14.13. The duration of this operation will vary depending on your server hardware and the number of node objects on your Chef Infra Server.
+
+#### Steps To Enable External OpenSearch
+1. Set the `opensearch['external']` attribute to `true`.
+2. Set the `opensearch['external_url']` attribute to the external OpenSearch URL.
+For example:
+```bash
+opensearch['external'] = true
+opensearch['external_url'] = "http://127.0.0.1:9200"
+```
```
Comment (IanMadd) [inline docs-chef-io/content/server/upgrades.md:60]: These need blank spaces around the heading, list items, and code. Also ordered list items should all be 1, Hugo will give them the right number when they're rendered.

````suggestion

1. Set the `opensearch['external']` attribute to `true`.
1. Set the `opensearch['external_url']` attribute to the external OpenSearch URL.

For example:

```bash
opensearch['external'] = true
opensearch['external_url'] = "http://127.0.0.1:9200"
```
````
